Request was for a topic or partition that does not exist on this broker

this error happens sometimes
> --- FAIL: TestKafkaConsumer_New_MockBroker (10.00s)
> 
>     errors.go:26: 
> 
>         	Error Trace:	errors.go:26
> 
>         	            				consumer_test.go:350
> 
>         	            				consumer_test.go:359
> 
>         	            				timeout.go:33
> 
>         	            				asm_amd64.s:1357
> 
>         	Error:      	Received unexpected error:
> 
>         	            	kafka server: Request was for a topic or partition that does not exist on this broker.
> 
>         	Test:       	TestKafkaConsumer_New_MockBroker
> 
>     errors.go:30: kafka server: Request was for a topic or partition that does not exist on this broker.
> 
>     timeout.go:39: Test ran out of time

--- a/consumer/consumer_test.go
+++ b/consumer/consumer_test.go
@@ -331,6 +331,9 @@ func TestKafkaConsumer_ProcessMessage_OrganizationIsNotAllowed(t *testing.T) {
 	assert.EqualError(t, err, "organization ID is not whitelisted")
 }
 
+// newConsumerWithMockBroker creates new mock consumer with mock broker
+// don't forget to wrap a calling test to helpers.RunTestWithTimeout,
+// because it can wait for mock broker creation forever
 func newConsumerWithMockBroker(t *testing.T) (storage.Storage, *helpers.MockBroker, consumer.Consumer) {
 	mockStorage := helpers.MustGetMockStorage(t, false)
 	mockBroker := helpers.MustNewMockBroker(t)
@@ -340,14 +343,27 @@ func newConsumerWithMockBroker(t *testing.T) (storage.Storage, *helpers.MockBrok
 	saramaConfig.Version = sarama.V0_10_0_1
 	saramaConfig.Producer.Return.Successes = true
 
-	mockConsumer, err := consumer.NewWithSaramaConfig(broker.Configuration{
-		Address:      mockBroker.Address,
-		Topic:        mockBroker.TopicName,
-		Group:        mockBroker.Group,
-		Enabled:      true,
-		OrgWhitelist: nil,
-	}, mockStorage, saramaConfig, false)
-	helpers.FailOnError(t, err)
+	var (
+		mockConsumer consumer.Consumer
+		err          error
+	)
+
+	for {
+		mockConsumer, err = consumer.NewWithSaramaConfig(broker.Configuration{
+			Address:      mockBroker.Address,
+			Topic:        mockBroker.TopicName,
+			Group:        mockBroker.Group,
+			Enabled:      true,
+			OrgWhitelist: nil,
+		}, mockStorage, saramaConfig, false)
+		// wait for topic to be created
+		if kErr, ok := err.(sarama.KError); ok && kErr == sarama.ErrUnknownTopicOrPartition {
+			time.Sleep(500 * time.Millisecond)
+			continue
+		}
+		helpers.FailOnError(t, err)
+		break
+	}
 
 	assert.NotNil(t, mockConsumer)
 

--- a/tests/helpers/mock_broker.go
+++ b/tests/helpers/mock_broker.go
@@ -17,7 +17,6 @@ package helpers
 import (
 	"context"
 	"testing"
-	"time"
 
 	"github.com/nash-io/jocko/protocol"
 
@@ -105,9 +104,6 @@ func (mockBroker *MockBroker) createTopic(topicName string) error {
 			return protocol.Errs[topicErrCode.ErrorCode]
 		}
 	}
-
-	// TODO: wait for it to actually be created
-	time.Sleep(500 * time.Millisecond)
 
 	return nil
 }

